File: 백준/Silver/1780. 종이의 개수/종이의 개수.cc
#include <iostream>
#include <vector>
using namespace std;

int n;
int cnt[3] = {0,0,0};
vector<vector<int>> arr;
void cut_paper(int num, int x, int y){
    int check = arr[y][x];
    bool flag = false;
    for (int i = y; i < y + num; i++){
        for (int j = x; j < x + num; j++){
            if (arr[i][j] != check){
                flag = true;
                break;
            }       
        }
        if (flag) break;
    }
    if (flag) {
        int resize = num / 3;
        for (int i =0; i < 3; i++)
            for (int j =0; j < 3; j++)
                cut_paper(resize, x + (j * resize), y + (i * resize));
    }
    else cnt[check] ++;
}

int main(){
    cin >> n;
    int num;
    for (int i =0; i < n; i++){
        vector<int> temp;
        for (int j = 0; j < n; j++){
            cin >> num;
            temp.push_back(num+1);
        }
        arr.push_back(temp);
    }
    cut_paper(n, 0, 0);
    for (int i = 0; i < 3; i++) cout << cnt[i] << endl;
}
        
    
    

File: 백준/Silver/1780. 종이의 개수/종이의 개수.java
import java.util.Scanner;

public class Main {

	static int N;
	static int[] cnt = new int[3];
	static int[][] map;

	static int check(int y, int x, int len) {
		int n = map[y][x];
		for (int i = y; i < y + len; i++) {
			for (int j = x; j < x + len; j++) {
				if (map[i][j] != n)
					return -99;
			}
		}
		return n;
	}

	static void func(int y, int x, int len) {
		int n = check(y, x, len);
		if (n != -99) {
			cnt[n + 1]++;
			return;
		}
		len /= 3;
		for (int i = 0; i < 3; i++) {
			func(y + len * i, x, len);
			func(y + len * i, x + len, len);
			func(y + len * i, x + len * 2, len);
		}
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		map = new int[N][N];
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				map[i][j] = sc.nextInt();
			}
		}
		func(0, 0, N);
		for (int i = 0; i < 3; i++) {
			System.out.println(cnt[i]);
		}
	}
}